server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -1589,6 +1509,120 @@ static String resolveExpression(String expression, LongSupplier getTime) {
+                } else {
+                    throw aliasesNotSupportedException(expression);
+                }
+            }
+            if (indexAbstraction.isDataStreamRelated() && context.includeDataStreams() == false) {
+                if (ignoreUnavailable) {
+                    return false;
+                } else {
+                    IndexNotFoundException infe = notFoundException(expression);
+                    // Allows callers to handle IndexNotFoundException differently based on whether data streams were excluded.
+                    infe.addMetadata(EXCLUDED_DATA_STREAMS_KEY, "true");
+                    throw infe;
+                }
+            }
+            return true;
+        }
+
+        private static void validateAliasOrIndex(ExpressionList.Expression expression) {
+            if (Strings.isEmpty(expression.expression())) {
+                throw notFoundException(expression.expression());
+            }
+            // Expressions can not start with an underscore. This is reserved for APIs. If the check gets here, the API
+            // does not exist and the path is interpreted as an expression. If the expression begins with an underscore,
+            // throw a specific error that is different from the [[IndexNotFoundException]], which is typically thrown
+            // if the expression can't be found.
+            if (expression.expression().charAt(0) == '_') {
+                throw new InvalidIndexNameException(expression.expression(), "must not start with '_'.");
+            }
+        }
+    }
+
+    /**
+     * Used to iterate expression lists and work out which expression item is a wildcard or an exclusion.
+     */
+    public static final class ExpressionList implements Iterable<ExpressionList.Expression> {
+        private final List<Expression> expressionsList;
+        private final boolean hasWildcard;
+
+        public record Expression(String expression, boolean isWildcard, boolean isExclusion) {
+            public String get() {
+                if (isExclusion()) {
+                    // drop the leading "-" if exclusion because it is easier for callers to handle it like this
+                    return expression().substring(1);
+                } else {
+                    return expression();
+                }
+            }
+        }
+
+        /**
+         * Creates the expression iterable that can be used to easily check which expression item is a wildcard or an exclusion (or both).
+         * The {@param context} is used to check if wildcards ought to be considered or not.
+         */
+        public ExpressionList(Context context, List<String> expressionStrings) {
+            List<Expression> expressionsList = new ArrayList<>(expressionStrings.size());
+            boolean wildcardSeen = false;
+            for (String expressionString : expressionStrings) {
+                boolean isExclusion = expressionString.startsWith("-") && wildcardSeen;
```

Comment:
Is it expected that "-foo,bar*" will say that "-foo" is not an exclusion ? (but "bar*,-foo" it is an exclusion).  I remember some conversation about this in the past but wanted to ensure this is intentional and passive.

Reply:
Yes, this is intentional. It's one of those behaviors that haven't bothered me enough to raise the discussion for change.